### PR TITLE
[Ember-Migration] Added local storage functionality and API Integration for Join Section

### LINKED ISSUE #389: [Ember-Migration] Add local storage functionality and API Integration for Join Section
## What to do?
- [ ] Add the functionality to store the data of all steps in local storage
- [ ] Ensure to get the latest data on jumping to previous step
- [ ] Retrieve all the data at the second last step and call `/users/self/intro` API
- [ ] Fill the `p` tag with `https://api.realdevsquad.com/users/<USERID>/intro`

## Data (Key Value) 
image

## Doubt Points
- Should we clear the local storage?
- Should the user be allowed to fill this only single time?
- What is the scenario if user want to edit their response?

### Possible Solution for above points
- Do not clear the local storage cache
- No they should be allowed to edit their response.
- There should be a option to go to the Step-1 of the form on the last page, so that users can edit their responses.

### PR DESCRIPTION
**Pull request by satyam73**
### Issue
Closes #389 

### What is the change?

I have added local storage and api functionality for the join section by saving all the form data in local storage and combined all the data at last step which will be used to send the api request.

API Endpoint - https://api.realdevsquad.com/users/self/intro

### Is Development Tested?
- [x] Yes

### Before / After Change Screenshots
https://user-images.githubusercontent.com/85296770/215271366-6d8401d9-9d0f-445b-bc9c-18adce0b9fd8.mp4

### CODE REVIEW

**Comment by Maheima** [comment]
Please update the description keeping in mind the points written at the end of it

> Note :
What to delete and what not to

> Keep the headings and answer the questions.
Delete the examples
Delete the description below headings What is the change? & Before/After Screenshots

**Comment by satyam73** [comment]
> Please update the description keeping in mind the points written at the end of it
> 
> > Note :
> > What to delete and what not to
> 
> > Keep the headings and answer the questions.
> > Delete the examples
> > Delete the description below headings What is the change? & Before/After Screenshots

Yeah deleted the examples section👍

File: app/components/join-steps/step-five.js [hunk 1]
```diff
@@ -5,11 +5,10 @@ import { tracked } from '@glimmer/tracking';
 import { TOAST_OPTIONS } from '../../constants/toast-options';
 import { ANKUSH_TWITTER } from '../../constants/urls';
 export default class StepFiveComponent extends Component {
-  ANKUSH_TWITTER = ANKUSH_TWITTER;
-
   @service toast;
-  @tracked joinLink =
-    'https://api.realdevsquad.com/users/1butZN1HmIr0rwQmaVXp/intro';
+  ANKUSH_TWITTER = ANKUSH_TWITTER;
+  @tracked id = localStorage.getItem('id');
```
Comment (rohan09-raj) [inline app/components/join-steps/step-five.js:10]: This can be `userId`

File: app/components/join-steps/step-four.js [hunk 2]
```diff
@@ -1,66 +1,101 @@
 import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
 
 export default class StepFourComponent extends Component {
+  @tracked stepOneData = JSON.parse(localStorage.getItem('stepOneData'));
+  @tracked stepTwoData = JSON.parse(localStorage.getItem('stepTwoData'));
+  @tracked stepThreeData = JSON.parse(localStorage.getItem('stepThreeData'));
+  @tracked allStepsData = {};
+  @tracked isLoading = this.login.isLoading;
```
Comment (rohan09-raj) [inline app/components/join-steps/step-four.js:9]: Where the service is being imported ?

File: app/components/join-steps/step-three.js [hunk 3]
```diff
@@ -1,10 +1,38 @@
...
 import { heardFrom } from '../../constants/social-data';
 import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+import { validator } from '../../utils/validator';
 
 export default class StepThreeComponent extends Component {
-  @tracked reasonToJoin = '';
-  @tracked foundFrom = '';
-
+  // @tracked foundFrom = '';
```
Comment (rohan09-raj) [inline app/components/join-steps/step-three.js:8]: Can we please remove the commented code?

File: app/utils/validator.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+export const validator = (value, words, length = 1) => {
```
Comment (rohan09-raj) [inline app/utils/validator.js:1]: What is the use of `length` here?

File: config/environment.js [hunk 5]
```diff
@@ -19,7 +19,8 @@ module.exports = function (environment) {
     },
   };
 
-  ENV.BASE_API_URL = 'https://api.realdevsquad.com';
+  // ENV.BASE_API_URL = 'https://api.realdevsquad.com';
+  ENV.BASE_API_URL = 'http://localhost:3000';
```
Comment (rohan09-raj) [inline config/environment.js:23]: Can we please revert these changes?

File: app/constants/urls.js [hunk 6]
```diff
@@ -20,6 +20,10 @@ export const AUTH = {
   SIGN_UP: 'https://my.realdevsquad.com/signup',
 };
 
+export const API_URL = {
+  JOIN_URL: 'https://api.realdevsquad.com/users/self/intro',
+};
```
Comment (rohan09-raj) [inline app/constants/urls.js:25]: Make a new constants file, for API URLs named as `APIs.js` and import the BASE_URL from the `environment.js` then use it something like this
```JS
import ENV from 'config/environments.js`

BASE_API_URL = ENV.BASE_API_URL
const JOIN_URL = `${BASE_API_URL}/users/self/intro`
```

The code is not exactly same as in the working state, it's there for reference only

File: tests/integration/components/dropdown-test.js [hunk 7]
```diff
@@ -1,29 +1,34 @@
...
-      @placeHolder={{this.placeholder}}
-      @value={{this.value}}
-      @options={{this.options}}
+    @name={{this.name}}
+    @field={{this.field}}
+    @required={{this.required}}
+    @placeHolder={{this.placeholder}}
+    @value={{this.value}}
+    @options={{this.options}}
+    @onChange={{this.onChange}}
```
Comment (rohan09-raj) [inline tests/integration/components/dropdown-test.js:31]: Can we please have 2 tab spaces for props?

File: tests/integration/components/step-one-test.js [hunk 8]
```diff
@@ -7,9 +7,22 @@ module('Integration | Component | step-one', function (hooks) {
...
+    });
+    this.set('setIsValid', () => {
+      assert.ok(true, 'setIsValid works!');
+    });
+
+    await render(hbs`
+    <JoinSteps::StepOne 
+    @setIsPreValid={{this.setIsPreValid}}
+    @isValid={{this.isValid}}
+    @setIsValid={{this.setIsValid}}
```
Comment (rohan09-raj) [inline tests/integration/components/step-one-test.js:24]: Can we please have 2 tab spacing here?

File: tests/integration/components/step-three-test.js [hunk 9]
```diff
@@ -7,9 +7,22 @@ module('Integration | Component | step-three', function (hooks) {
...
+    });
+    this.set('setIsValid', () => {
+      assert.ok(true, 'setIsValid works!');
+    });
+
+    await render(hbs`
+    <JoinSteps::StepThree
+    @setIsPreValid={{this.setIsPreValid}}
+    @isValid={{this.isValid}}
+    @setIsValid={{this.setIsValid}}
```
Comment (rohan09-raj) [inline tests/integration/components/step-three-test.js:24]: Can we please have 2 tab spacing here?

File: tests/integration/components/step-two-test.js [hunk 10]
```diff
@@ -7,9 +7,22 @@ module('Integration | Component | step-two', function (hooks) {
...
+    });
+    this.set('setIsValid', () => {
+      assert.ok(true, 'setIsValid works!');
+    });
+
+    await render(hbs`
+    <JoinSteps::StepTwo
+    @setIsPreValid={{this.setIsPreValid}}
+    @isValid={{this.isValid}}
+    @setIsValid={{this.setIsValid}}
```
Comment (rohan09-raj) [inline tests/integration/components/step-two-test.js:24]: Can we please have 2 tab spacing here?

**Review by rohan09-raj — CHANGES_REQUESTED** [review]

File: app/utils/validator.js (hunk 4 above)
Comment (satyam73) [reply app/utils/validator.js:1]: There is some validation on backend for that purpose I used length too with default value.

File: app/components/stepper.js [hunk 11]
```diff
@@ -2,21 +2,92 @@ import Component from '@glimmer/component';
 import { tracked } from '@glimmer/tracking';
 import { action } from '@ember/object';
 import { TITLE_MESSAGES } from '../constants/stepper-data';
+import { inject as service } from '@ember/service';
+import { TOAST_OPTIONS } from '../constants/toast-options';
+// import { API_URL } from '../constants/urls';
```
Comment (rohan09-raj) [inline app/components/stepper.js:7]: Can we please remove the commented code ?

File: app/utils/validator.js (hunk 4 above)
Comment (rohan09-raj) [reply app/utils/validator.js:1]: Can you please point me to that ?

File: app/components/join-steps/step-five.js [hunk 12]
```diff
@@ -5,11 +5,10 @@ import { tracked } from '@glimmer/tracking';
...
 import { ANKUSH_TWITTER } from '../../constants/urls';
 export default class StepFiveComponent extends Component {
-  ANKUSH_TWITTER = ANKUSH_TWITTER;
-
   @service toast;
-  @tracked joinLink =
-    'https://api.realdevsquad.com/users/1butZN1HmIr0rwQmaVXp/intro';
+  ANKUSH_TWITTER = ANKUSH_TWITTER;
+  @tracked userId = localStorage.getItem('id');
+  @tracked joinLink = `https://api.realdevsquad.com/users/${this.userId}/intro`;
```
Comment (rohan09-raj) [inline app/components/join-steps/step-five.js:11]: Please use `apis.js` constants for this as well.

File: app/utils/validator.js (hunk 4 above)
Comment (satyam73) [reply app/utils/validator.js:1]: This is the validation done on the backend API
image

**Review by rohan09-raj — APPROVED** [review]
Great work 💯 , looks good to me.
